### PR TITLE
aflplusplus: 4.10c -> 4.20c

### PR DESCRIPTION
**Pull request by Sigmanificient**
## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by Mindavi — DISMISSED** [review]
Hum, that's not gonna work :). Gonna need at least an update of the hash as well. And upstream has updated their customized qemu, so I prefer to keep in sync with that too.

I started making the update but haven't found time to finalise it yet, would be great if you could have a look 🙂.

**Comment by Sigmanificient** [comment]
> Hum, that's not gonna work :). Gonna need at least an update of the hash as well. And upstream has updated their customized qemu, so I prefer to keep in sync with that too.
> 
> I started making the update but haven't found time to finalise it yet, would be great if you could have a look 🙂.

Oh, I  completely forgot to update that hash, not sure why it built!

**Review by nicoonoclaste — DISMISSED** [review]

**Comment by Mindavi** [comment]
Qemu should still be updated to the latest version upstream uses.

**Comment by Sigmanificient** [comment]
> > Oh, I completely forgot to update that hash, not sure why it built!
> 
> Since the hash was the same, nix reused the old version's source. There are two things you could do to prevent that sort of issue:
> 
>     * add a test with [`testVersion`](https://nixos.org/manual/nixpkgs/unstable/#tester-testVersion) ;
> 
>     * use an [`updateScript`](https://nixos.org/manual/nixpkgs/unstable/#var-passthru-updateScript) instead of updating manually.
> 
> 
> @Sigmanificient, would you want to do that yourself, or should I do it prior to merging?

May you go ahead, I'm not fully sure how you want to do it

**Comment by SuperSandro2000** [comment]
Fails to build:


```
make: Leaving directory '/build/source'
[*] 12 test cases completed.
[-] not all test cases were executed
[!] failure in tests :-(
```

**Comment by doronbehar** [comment]
@Sigmanificient would you like to keep working on this? I see 4.21c	was released 3 days ago:
https://github.com/AFLplusplus/AFLplusplus/releases

**Comment by Sigmanificient** [comment]
> @Sigmanificient would you like to keep working on this? I see 4.21c was released 3 days ago: [AFLplusplus/AFLplusplus/releases](https://github.com/AFLplusplus/AFLplusplus/releases)

```

+++ Testing aborted programmatically +++
[!] Time limit was reached
[*] Statistics: 0 new corpus items found, 0.11% coverage achieved, 0 crashes saved, 0 timeouts saved, total runtime 0 days, 0 hrs, 0 min, 7 sec
[+] We're done here. Have a nice day!

CUT------------------------------------------------------------------CUT
[!] afl-fuzz is not working correctly with qemu_mode cmplog
[*] running afl-fuzz for persistent qemu_mode, this will take appro
```
Tried to build with `4.21c` no sure what to do about this

**Comment by doronbehar** [comment]
That is indeed weird. If you don't find the cause for this failure we can disable the tests in the worse case, with a comment explaining this situation. If you want to continue debug this, I'd try to do so in a `nix develop` environment in a git checkout.

**Comment by Sigmanificient** [comment]
I clearly underestimated the difficulty of this package when i opened that pr quite some time ago  :sweat_smile:
Feel like someone else would do a much better job, especially since i dont have much free time currently

**Comment by doronbehar** [comment]
Yes it would have been nice if the maintainers would have responded.. If you are not sure whether you are breaking something with the update, and you haven't tested it properly, it's legitimate to let it go.

**Comment by Mindavi** [comment]
I'll probably pick this up myself soon-ish, but I'm generally a bit conservative with updating this since the upstream build system is not very nixpkgs-friendly. But I'll definitely pick this whenever I do so, so thanks for that.

**Comment by Sigmanificient** [comment]
This does build

**Comment by Sigmanificient** [comment]
I thought it was all good,  but tests are still failling sadly :<

**Comment by Mindavi** [comment]
I don't really understand, it seemed to work fine locally but maybe I mislooked... The unit tests are less interesting anyway though.

**Comment by doronbehar** [comment]
You may want to try to rebase the PR branch on top of latest `master` to reproduce the failure.

**Comment by Mindavi** [comment]
Hmm, even after a rebase on master:

```
aflplusplus> [  PASSED  ] 1 test(s).
aflplusplus> make: Leaving directory '/build/source'
aflplusplus> [*] 12 test cases completed.
aflplusplus> [-] not all test cases were executed
aflplusplus> [+] all tests were successful :-)
aflplusplus> installCheckPhase completed in 3 minutes 7 seconds
```

Let's have another attempt.

@ofborg build aflplusplus

**Comment by doronbehar** [comment]
Hmm still fails...

**Comment by Mindavi** [comment]
This build seems to have succeeded: https://github.com/NixOS/nixpkgs/pull/305773/checks?check_run_id=27428395866.

**Comment by doronbehar** [comment]
> This build seems to have succeeded: [#305773 (checks)](https://github.com/NixOS/nixpkgs/pull/305773/checks?check_run_id=27428395866).

Right! What I saw was an old attempt probably. You edited your previous comment and I think ofborg didn't catch the change `eval` -> `build alfplusplus`... Never mind, it seems to be OK now.

File: pkgs/tools/security/aflplusplus/qemu.nix
Comment (doronbehar) [inline pkgs/tools/security/aflplusplus/qemu.nix:27]: In the next update of AFLplusplus, how would we be able to know what revision to use? Is this a submodule or something alike?

**Review by doronbehar — COMMENTED** [review]
Looks good overall. Should be ready to merge once this small issue is fixed.

File: pkgs/tools/security/aflplusplus/qemu.nix
Comment (Sigmanificient) [reply pkgs/tools/security/aflplusplus/qemu.nix:27]: I think this is this: https://github.com/AFLplusplus/AFLplusplus/commits/stable/qemu_mode/QEMUAFL_VERSION

Comment (Sigmanificient) [reply pkgs/tools/security/aflplusplus/qemu.nix:27]: This commit https://github.com/AFLplusplus/AFLplusplus/commit/07bc202e0ad940e0cc7c8770f69ceb32ed851384 sets it to `40033af`

Comment (doronbehar) [reply pkgs/tools/security/aflplusplus/qemu.nix:27]: > I think this is this: [AFLplusplus/AFLplusplus@`stable`/qemu_mode/QEMUAFL_VERSION (commits)](https://github.com/AFLplusplus/AFLplusplus/commits/stable/qemu_mode/QEMUAFL_VERSION)

@Sigmanificient this is a great link to put in a comment near that revision. Could you please do that? Besides that this PR can be merged IMO.

**Comment by Sigmanificient** [comment]
@doronbehar I noticed a depreciated `replace` (substitueInPlace) while building, may I update them?

**Comment by doronbehar** [comment]
> @doronbehar I noticed a depreciated `replace` (substitueInPlace) while building, may I modernize the derivation a bit?

Yea sure go ahead. I'd recommend do so in a separate commit, and before the update commit.

**Comment by Sigmanificient** [comment]
One of the changes is `sha` -> `hash`, so i keep the commit after to avoid the conflict

**Review by doronbehar — COMMENTED** [review]
If you are attempting to modernize this, how about go all the way? :)

Use `nixfmt` (from package `nixfmt-rfc-style`), and use `finalAttrs` (function argument to `mkDerivation`). You can put each step in a dedicated PR.

Bonus: https://github.com/NixOS/nixpkgs/issues/208242 .

**Comment by Sigmanificient** [comment]
Well tough that running `nixfmt` would ofuscate too much the diff

**Comment by Sigmanificient** [comment]
Result of `nixpkgs-review pr 305773` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>aflplusplus</li>
    <li>libdislocator</li>
  </ul>
</details>

**Comment by Sigmanificient** [comment]
Applied formatting and fixed #208242 
I will not try to migrate it to by-name though :sweat_smile:

**Review by Mindavi — APPROVED** [review]
Let's see what ofborg thinks now. Thanks for modernizing this package description<3

**Comment by Sigmanificient** [comment]
Result of `nixpkgs-review pr 305773` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>aflplusplus</li>
    <li>libdislocator</li>
  </ul>
</details>

**Pull request merged**
